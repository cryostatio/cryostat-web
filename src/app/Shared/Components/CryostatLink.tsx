/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { BASEPATH, toPath } from '@app/utils/utils';
import React from 'react';
import { Link, LinkProps, Path, To } from 'react-router-dom-v5-compat';

export interface CryostatLinkProps extends LinkProps {}

/**
 * Formats a To (string | Partial\<Path\>) by prepending a basepath if necessary
 * @param {string | Partial<Path>} destination - the target destination
 */
const toDestination = (destination: To) => {
  if (BASEPATH) {
    if (typeof destination === 'string') {
      return toPath(destination);
    } else {
      (destination as Partial<Path>).pathname = `/${BASEPATH}${(destination as Partial<Path>).pathname}`;
      return destination as Partial<Path>;
    }
  } else {
    return destination;
  }
};

export const CryostatLink: React.FC<CryostatLinkProps> = ({ to, onClick, ...props }) => {
  return <Link to={toDestination(to)} onClick={onClick} {...props}></Link>;
};
