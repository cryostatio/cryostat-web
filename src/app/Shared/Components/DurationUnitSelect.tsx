/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { MenuToggle, MenuToggleElement, Select, SelectList, SelectOption } from '@patternfly/react-core';
import { TFunction } from 'i18next';
import * as React from 'react';
import { useTranslation } from 'react-i18next';

export enum DurationUnit {
  HOUR = 60 * 60,
  MINUTE = 60,
  SECOND = 1,
}

const SupportedDurationUnits = Object.values(DurationUnit).filter((entry) => typeof entry === 'string');

export const getDurationUnitDisplay = (t: TFunction, unit: DurationUnit, compact?: boolean): string => {
  const suffix = compact ? 'compact' : 'other';
  switch (unit) {
    case DurationUnit.HOUR:
      return t(`HOUR_${suffix}`, { ns: 'common' });
    case DurationUnit.MINUTE:
      return t(`MINUTE_${suffix}`, { ns: 'common' });
    case DurationUnit.SECOND:
      return t(`SECOND_${suffix}`, { ns: 'common' });
    default:
      throw new Error(`Unknown unit enum value: ${unit}`);
  }
};

export interface DurationUnitSelectProps {
  onSelect?: (unit: DurationUnit) => void;
  selected: DurationUnit;
}

export const DurationUnitSelect: React.FC<DurationUnitSelectProps> = ({ onSelect, selected, ...props }) => {
  const { t } = useTranslation();
  const [open, setOpen] = React.useState(false);

  const handleToggle = React.useCallback(() => setOpen((open) => !open), [setOpen]);

  const handleSelect = React.useCallback(
    (e, value: DurationUnit) => {
      setOpen(false);
      onSelect && onSelect(value);
    },
    [onSelect, setOpen],
  );

  const options = React.useMemo(
    () =>
      SupportedDurationUnits.map((unit) => (
        <SelectOption key={unit} value={DurationUnit[unit]}>
          {getDurationUnitDisplay(t, DurationUnit[unit])}
        </SelectOption>
      )),
    [t],
  );

  const toggle = React.useCallback(
    (toggleRef: React.Ref<MenuToggleElement>) => (
      <MenuToggle ref={toggleRef} onClick={handleToggle} isExpanded={open}>
        {getDurationUnitDisplay(t, selected)}
      </MenuToggle>
    ),
    [handleToggle, open, t, selected],
  );

  return (
    <Select
      {...props}
      toggle={toggle}
      isOpen={open}
      selected={selected}
      onSelect={handleSelect}
      onOpenChange={setOpen}
      onOpenChangeKeys={['Escape']}
    >
      <SelectList>{options}</SelectList>
    </Select>
  );
};
