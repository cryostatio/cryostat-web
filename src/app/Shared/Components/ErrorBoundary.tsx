/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import build from '@app/build.json';
import { useSubscriptions } from '@app/utils/hooks/useSubscriptions';
import {
  Bullseye,
  EmptyState,
  EmptyStateBody,
  EmptyStateIcon,
  EmptyStateVariant,
  Text,
  TextVariants,
  EmptyStateHeader,
} from '@patternfly/react-core';
import { ExclamationCircleIcon } from '@patternfly/react-icons';
import * as React from 'react';
import { Trans, useTranslation } from 'react-i18next';
import { ServiceContext } from '../Services/Services';

export interface ErrorBoundaryProps {
  renderFallback: (error: Error) => React.ReactNode;
  children?: React.ReactNode;
}

export interface ErrorBoundaryState {
  error?: Error;
}

export class ErrorBoundary extends React.Component<ErrorBoundaryProps, ErrorBoundaryState> {
  constructor(props: ErrorBoundaryProps) {
    super(props);
    this.state = { error: undefined };
  }

  static getDerivedStateFromError(error: Error): ErrorBoundaryState {
    // Update state so the next render will show the fallback UI.
    return { error };
  }

  componentDidCatch(error: Error, info: React.ErrorInfo) {
    console.warn(error, info.componentStack);
  }

  render(): React.ReactNode {
    if (this.state.error) {
      return this.props.renderFallback(this.state.error);
    }
    return this.props.children;
  }
}

export interface DefaultFallBackProps {
  error?: Error;
}

export const DefaultFallBack: React.FC<DefaultFallBackProps> = ({ error, ...props }) => {
  const { t } = useTranslation();
  const addSubcription = useSubscriptions();
  const serviceContext = React.useContext(ServiceContext);
  const [cryostatVersion, setCryostatVersion] = React.useState(undefined as string | undefined);

  React.useEffect(() => {
    addSubcription(serviceContext.api.cryostatVersion().subscribe(setCryostatVersion));
  }, [serviceContext, setCryostatVersion, addSubcription]);

  return (
    <Bullseye {...props}>
      <EmptyState variant={EmptyStateVariant.lg}>
        <EmptyStateHeader
          titleText={<>{t('SOMETHING_WENT_WRONG', { ns: 'common' })}</>}
          icon={<EmptyStateIcon icon={ExclamationCircleIcon} color="red" />}
          headingLevel={'h1'}
        />
        <EmptyStateBody>
          <p>{t('ERROR_BOUNDARY.ERROR_MESSAGE', { message: error?.message || 'Unknown error.' })}</p>
          <Trans
            t={t}
            values={{
              knownIssue: t('AboutDescription.KNOWN_ISSUES'),
              fileReport: t('AboutDescription.FILE_A_REPORT'),
            }}
            components={{
              issue: <Text component={TextVariants.a} target="_blank" href={build.knownIssuesUrl} />,
              report: (
                <Text
                  component={TextVariants.a}
                  target="_blank"
                  href={build.fileIssueUrl.replace('__REPLACE_VERSION__', cryostatVersion || 'unknown')}
                />
              ),
            }}
          >
            ERROR_BOUNDARY.RESOLVE_MESSAGE
          </Trans>
        </EmptyStateBody>
      </EmptyState>
    </Bullseye>
  );
};
