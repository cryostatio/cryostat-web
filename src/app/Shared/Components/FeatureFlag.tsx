/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { FeatureLevel } from '@app/Shared/Services/service.types';
import { ServiceContext } from '@app/Shared/Services/Services';
import { useSubscriptions } from '@app/utils/hooks/useSubscriptions';
import * as React from 'react';

export interface DynamicFeatureFlagProps {
  levels: FeatureLevel[];
  component: (level: FeatureLevel) => React.ReactNode;
  defaultComponent?: React.ReactNode;
}

export const DynamicFeatureFlag: React.FC<DynamicFeatureFlagProps> = ({ levels, component, defaultComponent }) => {
  const context = React.useContext(ServiceContext);
  const addSubscription = useSubscriptions();
  const [activeLevel, setActiveLevel] = React.useState(FeatureLevel.PRODUCTION);

  React.useLayoutEffect(() => {
    addSubscription(context.settings.featureLevel().subscribe((featureLevel) => setActiveLevel(featureLevel)));
  }, [addSubscription, context.settings, setActiveLevel]);

  const toRender = React.useMemo(() => {
    if (levels.includes(activeLevel)) {
      return component(activeLevel);
    }
    return defaultComponent;
  }, [levels, activeLevel, component, defaultComponent]);

  return <>{toRender}</>;
};

export interface FeatureFlagProps {
  strict?: boolean;
  level: FeatureLevel;
  children?: React.ReactNode | undefined;
}

export const FeatureFlag: React.FC<FeatureFlagProps> = ({ level, strict, children }) => {
  const levels = React.useMemo(
    () => (strict ? [level] : [...Array.from({ length: level + 1 }, (_, i) => i)]),
    [strict, level],
  );
  const component = React.useCallback((_: FeatureLevel) => <>{children}</>, [children]);

  return <DynamicFeatureFlag levels={levels} component={component} />;
};
