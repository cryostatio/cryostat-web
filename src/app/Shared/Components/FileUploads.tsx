/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { CancelUploadModal } from '@app/Modal/CancelUploadModal';
import { NotificationsContext } from '@app/Shared/Services/Notifications.service';
import {
  MultipleFileUpload,
  MultipleFileUploadMain,
  MultipleFileUploadStatus,
  MultipleFileUploadStatusItem,
} from '@patternfly/react-core';
import { InProgressIcon, UploadIcon } from '@patternfly/react-icons';
import * as React from 'react';
import { Subject } from 'rxjs';
import { DropzoneAccept, FileRejection } from './types';

export type ProgressVariant = 'success' | 'danger' | 'warning';

export interface FUpload {
  file: File;
  abortSignal: Subject<void>;
  progress?: {
    progressValue: number;
    progressVariant: ProgressVariant;
  };
  helperText?: React.ReactNode;
  error?: Error; // error.message take precedence over helperText
}

export interface UploadCallbacks {
  getProgressUpdateCallback: (filename: string) => (progress: number | string) => void;
  onSingleSuccess: (filename: string, message?: React.ReactNode) => void;
  onSingleFailure: (filename: string, error: Error) => void;
}

export interface MultiFileUploadProps {
  submitRef?: React.RefObject<HTMLDivElement>;
  abortRef?: React.RefObject<HTMLDivElement>;
  uploading: boolean;
  displayAccepts: string[];
  dropZoneAccepts: DropzoneAccept;
  onFilesChange?: (files: FUpload[]) => void;
  onFileSubmit: (fileUploads: FUpload[], uploadCallbacks: UploadCallbacks) => void;
  titleIcon?: React.ReactNode;
  titleText?: React.ReactNode;
  titleTextSeparator?: React.ReactNode;
  infoText?: React.ReactNode;
}

export const MultiFileUpload: React.FC<MultiFileUploadProps> = ({
  titleIcon,
  titleText,
  titleTextSeparator,
  infoText,
  displayAccepts,
  dropZoneAccepts,
  onFileSubmit,
  uploading,
  onFilesChange,
  submitRef,
  abortRef,
}) => {
  const notifications = React.useContext(NotificationsContext);
  const [fileUploads, setFileUploads] = React.useState<FUpload[]>([]);
  const [showCancelPrompt, setShowCancelPrompt] = React.useState(false);

  const handleFileDrop = React.useCallback(
    (_, droppedFiles: File[]) => {
      setFileUploads((old) => {
        // Check for re-uploads
        const currentFilenames = old.map((fileUpload) => fileUpload.file.name);
        const reUploadFilenames = droppedFiles.filter((f) => currentFilenames.includes(f.name)).map((f) => f.name);

        const newFileUploads = [
          ...old.filter((fileUpload) => !reUploadFilenames.includes(fileUpload.file.name)),
          ...droppedFiles.map(
            (f) =>
              ({
                file: f,
                abortSignal: new Subject<void>(),
              }) as FUpload,
          ),
        ];
        onFilesChange && onFilesChange(newFileUploads);
        return newFileUploads;
      });
    },
    [setFileUploads, onFilesChange],
  );

  const handleFileReject = React.useCallback(
    (fileRejections: FileRejection[]) => {
      fileRejections.forEach(({ file, errors }) => {
        errors.forEach(({ message }) => {
          notifications.warning(`Rejected file: ${file.name}`, message);
        });
      });
    },
    [notifications],
  );

  const handleFileRemove = React.useCallback(
    (removedFilename: string, abort = false) => {
      if (abort) {
        const fileToAbort = fileUploads.find((fileUpload) => fileUpload.file.name === removedFilename);
        fileToAbort?.abortSignal.next();
      } else {
        setFileUploads((old) => {
          const newFileUploads = old.filter((fileUpload) => fileUpload.file.name !== removedFilename);
          onFilesChange && onFilesChange(newFileUploads);
          return newFileUploads;
        });
      }
    },
    [fileUploads, setFileUploads, onFilesChange],
  );

  const getProgressUpdateCallback = React.useCallback(
    (filename: string) => {
      return (progress: string | number) => {
        setFileUploads((old) => {
          const match = old.find((f) => f.file.name === filename);
          if (match) {
            return [
              ...old.filter((f) => f.file.name !== filename),
              {
                ...match,
                progress: {
                  progressValue: progress,
                  progressVariant: `${progress}` == '100' ? 'success' : undefined,
                },
              } as FUpload,
            ];
          }
          return old;
        });
      };
    },
    [setFileUploads],
  );

  const onSingleFailure = React.useCallback(
    (filename: string, error: Error) => {
      setFileUploads((old) => {
        const match = old.find((f) => f.file.name === filename);
        if (match) {
          return [
            ...old.filter((f) => f.file.name !== filename),
            {
              ...match,
              progress: {
                progressValue: match.progress?.progressValue || 0,
                progressVariant: 'danger',
              },
              error: error,
            } as FUpload,
          ];
        }
        return old;
      });
    },
    [setFileUploads],
  );

  const onSingleSuccess = React.useCallback(
    (filename: string, message?: React.ReactNode) => {
      setFileUploads((old) => {
        const match = old.find((f) => f.file.name === filename);
        if (match) {
          return [
            ...old.filter((f) => f.file.name !== filename),
            {
              ...match,
              progress: {
                progressValue: 100,
                progressVariant: 'success',
              },
              helperText: message,
            } as FUpload,
          ];
        }
        return old;
      });
    },
    [setFileUploads],
  );

  const handleSubmit = React.useCallback(() => {
    setFileUploads((old) => {
      const toUploads = old.filter((f) => f.error || !f.progress); // Failed or newly uploaded
      onFileSubmit(toUploads, { getProgressUpdateCallback, onSingleSuccess, onSingleFailure });

      const newFileUploads = old.map(
        (f) =>
          ({
            ...f,
            error: undefined,
            progress: f.progress?.progressVariant === 'success' ? f.progress : undefined,
            helperText: f.progress?.progressVariant === 'success' ? 'Already uploaded' : undefined,
          }) as FUpload,
      );
      return newFileUploads;
    });
  }, [onFileSubmit, getProgressUpdateCallback, onSingleSuccess, onSingleFailure]);

  const handleCloseCancelModal = React.useCallback(() => setShowCancelPrompt(false), [setShowCancelPrompt]);

  const handleOpenCancelModal = React.useCallback(() => setShowCancelPrompt(true), [setShowCancelPrompt]);

  const handleAbortAll = React.useCallback(() => {
    fileUploads
      .filter((fileUpload) => !fileUpload.progress?.progressVariant)
      .map((fileUpload) => {
        fileUpload.abortSignal.next(); // trigger abort
      });

    handleCloseCancelModal();
  }, [fileUploads, handleCloseCancelModal]);

  const handleCancel = React.useCallback(() => {
    if (uploading) {
      handleOpenCancelModal();
    }
  }, [handleOpenCancelModal, uploading]);

  const numOfSuccessUploads = React.useMemo(() => {
    return fileUploads.filter((fUpload) => fUpload.progress && fUpload.progress.progressVariant === 'success').length;
  }, [fileUploads]);

  const sortedFileUploads = React.useMemo(() => {
    return fileUploads.sort((a, b) => a.file.name.localeCompare(b.file.name, undefined, { numeric: true }));
  }, [fileUploads]);

  return (
    <>
      {/* 
        TODO: Add back when supported 
        Reference: https://reactrouter.com/en/main/upgrading/v5#prompt-is-not-currently-supported
        <Prompt when={uploading} message="Are you sure you wish to cancel the file upload?" /> 
      */}
      <CancelUploadModal
        visible={showCancelPrompt}
        title="Upload in Progress"
        message="Are you sure you wish to cancel the file uploads? Successfully uploaded files won't be aborted."
        onYes={handleAbortAll}
        onNo={handleCloseCancelModal}
      />
      <MultipleFileUpload
        onFileDrop={handleFileDrop}
        dropzoneProps={{
          accept: dropZoneAccepts,
          onDropRejected: handleFileReject,
        }}
        disabled={uploading}
      >
        <MultipleFileUploadMain
          titleIcon={titleIcon || <UploadIcon />}
          titleText={titleText || 'Drag a file here'}
          titleTextSeparator={titleTextSeparator || 'or'}
          infoText={infoText || `Accepted file types: ${displayAccepts.join(', ')}`}
        />
        {fileUploads.length ? (
          <MultipleFileUploadStatus
            statusToggleText={`${numOfSuccessUploads} of ${fileUploads.length} files uploaded`}
            statusToggleIcon={<InProgressIcon />}
          >
            {sortedFileUploads.map((fileUpload) => (
              <MultipleFileUploadStatusItem
                file={fileUpload.file}
                key={fileUpload.file.name}
                onClearClick={() => handleFileRemove(fileUpload.file.name, uploading)}
                customFileHandler={(_) => undefined} // To disable built-in file reader and default styling
                progressValue={fileUpload.progress?.progressValue}
                progressVariant={fileUpload.progress?.progressVariant}
                progressHelperText={fileUpload.error?.message || fileUpload.helperText}
              />
            ))}
          </MultipleFileUploadStatus>
        ) : undefined}
      </MultipleFileUpload>
      {/* fake action buttons */}
      <div ref={submitRef} id={'start-upload-files'} hidden onClick={handleSubmit} />
      <div ref={abortRef} id={'abort-upload-files'} hidden onClick={handleCancel} />
    </>
  );
};
