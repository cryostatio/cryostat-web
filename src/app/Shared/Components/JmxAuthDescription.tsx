/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Text, TextContent, TextList, TextListItem, TextVariants } from '@patternfly/react-core';
import * as React from 'react';
import { DescriptionProps } from './types';

export const JmxAuthDescription: React.FC<React.PropsWithChildren<DescriptionProps>> = ({ children }) => {
  return (
    <TextContent>
      {children}
      <Text component={TextVariants.p}>
        JVM applications can be configured to require clients, such as Cryostat, to pass a challenge based
        authentication before establishing a connection.
      </Text>
      <Text component={TextVariants.p}>
        Check the deployment configuration of your JVM application for system properties such as:
      </Text>
      <TextList>
        <TextListItem>
          <Text component={TextVariants.pre}>com.sun.management.jmxremote.authenticate</Text>
        </TextListItem>
        <TextListItem>
          <Text component={TextVariants.pre}>com.sun.management.jmxremote.password.file</Text>
        </TextListItem>
        <TextListItem>
          <Text component={TextVariants.pre}>com.sun.management.jmxremote.login.config</Text>
        </TextListItem>
      </TextList>
      <Text component={TextVariants.p}>
        These authentication credentials are stored in encrypted storage managed by the Cryostat backend and used for
        manually managing recordings and event templates on target JVMs, as well as for Automated Rules which run in the
        background and open unattended target connections.
      </Text>
    </TextContent>
  );
};
