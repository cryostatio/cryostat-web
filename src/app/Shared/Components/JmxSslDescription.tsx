/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Text, TextContent, TextList, TextListItem, TextVariants } from '@patternfly/react-core';
import * as React from 'react';
import { DescriptionProps } from './types';

export const JmxSslDescription: React.FC<React.PropsWithChildren<DescriptionProps>> = ({ children }) => {
  return (
    <TextContent>
      {children}
      <Text component={TextVariants.p}>
        JVM applications can be configured to present an SSL certificate for incoming JMX connections. Clients, such as
        Cryostat, should be configured to trust these certificates so that the origin and authenticity of the connection
        data can be verified.
      </Text>
      <Text component={TextVariants.p}>
        Check the deployment configuration of your JVM application for system properties such as:
      </Text>
      <TextList>
        <TextListItem>
          <Text component={TextVariants.pre}>javax.net.ssl.keyStore</Text>
        </TextListItem>
        <TextListItem>
          <Text component={TextVariants.pre}>javax.net.ssl.keyStorePassword</Text>
        </TextListItem>
        <TextListItem>
          <Text component={TextVariants.pre}>com.sun.management.jmxremote.ssl.need.client.auth</Text>
        </TextListItem>
        <TextListItem>
          <Text component={TextVariants.pre}>javax.net.ssl.trustStore</Text>
        </TextListItem>
        <TextListItem>
          <Text component={TextVariants.pre}>javax.net.ssl.trustStorePassword</Text>
        </TextListItem>
        <TextListItem>
          <Text component={TextVariants.pre}>com.sun.management.jmxremote.registry.ssl</Text>
        </TextListItem>
      </TextList>
    </TextContent>
  );
};
