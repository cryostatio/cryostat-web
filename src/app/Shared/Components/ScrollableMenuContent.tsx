/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Panel, PanelMain, PanelMainBody } from '@patternfly/react-core';
import * as React from 'react';

export interface ScrollableMenuContentProps {
  maxHeight?: string;
}

// Use case: Menu footer or search bar needs to be sticky.
export const ScrollableMenuContent: React.FC<ScrollableMenuContentProps> = ({ children, maxHeight }) => {
  return (
    <Panel isScrollable>
      <PanelMain maxHeight={maxHeight}>
        <PanelMainBody style={{ padding: 0 }}>{children}</PanelMainBody>
      </PanelMain>
    </Panel>
  );
};
