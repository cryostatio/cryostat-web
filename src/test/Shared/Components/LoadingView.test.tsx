/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { LoadingView } from '@app/Shared/Components/LoadingView';
import { cleanup, render, screen } from '@testing-library/react';
import renderer, { act } from 'react-test-renderer';

describe('<LoadingView />', () => {
  afterEach(cleanup);

  it('renders correctly', async () => {
    let tree;
    await act(async () => {
      tree = renderer.create(<LoadingView />);
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should show spinner and title', async () => {
    render(<LoadingView title="Progressing" />);

    const title = screen.getByText('Progressing');
    expect(title).toBeInTheDocument();
    expect(title).toBeVisible();

    const spinner = screen.getByRole('progressbar');
    expect(spinner).toBeInTheDocument();
    expect(spinner).toBeVisible();
  });
});
