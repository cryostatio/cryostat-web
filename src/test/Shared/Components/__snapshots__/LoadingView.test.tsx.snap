// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoadingView /> renders correctly 1`] = `
<div
  className="pf-v5-l-bullseye"
>
  <div
    className="pf-v5-c-empty-state"
  >
    <div
      className="pf-v5-c-empty-state__content"
    >
      <div
        className="pf-v5-c-empty-state__header"
      >
        <div
          className="pf-v5-c-empty-state__icon"
        >
          <svg
            aria-hidden={false}
            aria-label="Contents"
            aria-valuetext="Loading..."
            className="pf-v5-c-spinner pf-m-xl"
            role="progressbar"
            viewBox="0 0 100 100"
          >
            <circle
              className="pf-v5-c-spinner__path"
              cx="50"
              cy="50"
              fill="none"
              r="45"
            />
          </svg>
        </div>
        <div
          className="pf-v5-c-empty-state__title"
        >
          <h2
            className="pf-v5-c-empty-state__title-text"
          >
            Loading
          </h2>
        </div>
      </div>
    </div>
  </div>
</div>
`;
